### DISCUSSION TITLE
Access single item at startup

### DISCUSSION THREAD

**Discussion by batjoc** [mergesort/Boutique#23] [Q&A] [answered]
First of all thanks for creating this great library 🙏 

Similar to the question asked in https://github.com/mergesort/Boutique/discussions/17, I was looking into a way to access a single item at startup. My use case is the following: in my app, there are several objects stored (not many, probably 10 at most with some attributes). Now what I want to provide is that when a user opens the app, the last selected object is shown. However I'm running into the problem that `Store.items` is empty at startup. 

I noticed that the discussion was marked as solved, however I think it was mainly based on the speed improvements, not the `Store.items` being valid at startup right? Is this something you plan to build?

**Comment by mergesort** [comment, marked as answer]
Hey @batjoc! Apologies but not too sorry for the delay, I've been on vacation and wanted to really give myself a nice break. 🙂

You're right that the first access of `Store.items` will return `[]` on startup, that's something I want to improve but I'm honestly not sure how. The issue is that there has to be a default value, and since the initial query of all the persisted `items` has to take place on the `@MainActor`, there will be some non-zero initialization time when the `Task` is dispatched.

I do think I have some potential help for you though. I've been working hard on a 2.0 releases of Boutique and Bodega, and one of the improvements is the ability to have a `@StoredValue` type, similar to `@Store`. As the name implies it can store a single value which should be accessible at startup. You can see the code in this [commit](https://github.com/mergesort/Boutique/commit/1bf02872f4758ebb330f1ee34cc5962d1bc0b951) and even pull it into your project now if you'd like.

I can recommend using the 2.0 branch, I'm going to cut a preview release version in the next day or two. There are some minimal API breaking changes, for the purposes of creating a better overall API. The migration should be very small and I think you'll see it's not hard to implement, the reason it's a preview release and not an official release is only because is I want to document a migration path, new features, and build some walkthroughs to show people how to use Boutique and Bodega.

Hope that helps, if that works lemme know how it goes, and if not lemme know if I can help otherwise!

**Reply by AzSiAz** [reply]
Instead of a global unique id why not allow to set it in `init` ?
This would allow to retrieve any value already in a store and let people use their own specialized store if they want instead of having only one unique value per store path

EDIT: clarify what I mean by specialized store

**Reply by mergesort** [reply]
@AzSiaZ That's a good idea that I hadn't considered! Rather than passing in a `Store` I pushed up a version of `@StoredValue` that takes a key parameter alongside the `StorageEngine`. The reason I don't want to make people pass `Store`s in is because then they need to deal with figuring out how to create a `cacheIdentifier` keyPath mapping for their `Item`, but doing it this way allows you to not even think about any `Store` semantics since they're all taken care of under the hood for you.

How does [this](https://github.com/mergesort/Boutique/commit/099a8d57db791e0358ad93e9e408628bd33fdfdf) look to you?

**Reply by AzSiAz** [reply]
Look fine but is there still a need for `UniqueItem` and would it work without a decode error ?

By the way does `Store`/`@Stored` have some form of instance uniqueness or is it a Bodega `StorageEngine` responsibility ?
My worry is that using multiple `@StoredValue` will load every value in store every time it's used which could lead to memory explosion in very large `Store` which are now a possibility considering sqlite usage by default or custom engine enabled by Bodega 2.0

**Reply by batjoc** [reply]
Thanks I guess this is indeed what I'm looking for. I was trying it out, but as @AzSiAz mentioned, when running this code I get a decode error: `keyNotFound(CodingKeys(stringValue: "value", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"value\", intValue: nil) (\"value\").", underlyingError: nil))`

This error happens as soon as I call this code: `try await self.$storedItem.set(item)`. Well technically not as soon, it happens when I restart the app and it runs this code `self.items = try await self.storageEngine.readAllData().map({ try decoder.decode(Item.self, from: $0) })`.

**Reply by mergesort** [reply]
I'll write more about this when I get home and can provide a better response, but after some feedback about the intention of `@StoredValue` I plan on reverting back to [this commit](https://github.com/mergesort/Boutique/commit/099a8d57db791e0358ad93e9e408628bd33fdfdf). The second version is closer to creating a Store scoped to returning one `Item`, which wasn't my original intention.

@batjoc Since I'm not at a computer I'm curious, does the key-less variant cause any problems for you? If not I'll iterate on the ergonomics a bit when I get home and continue down that path to solve the problem you're encountering.

**Reply by batjoc** [reply]
Yes both variants cause the same `keyNotFound` error, I thought I was making a mistake in my code, but after digging a bit deeper I found out it that it only happened when I set the `@StoredValue`. I tried to solve it, but as I'm just getting started with Swift after a long time using ObjectiveC I still have some trouble understanding all the magic 😉

**Reply by AzSiAz** [reply]
The keyless version is probably doing the same error because you already have data inside this `Store`

**Reply by mergesort** [reply]
I think @AzSiAz is right, could you try it on a fresh Store? I haven't run into that error in my tests and am wondering if it's a matter of the state being corrupted from my in-flight changes.

**Reply by batjoc** [reply]
@mergesort I tested it, but it also happens on a fresh Store. As I said, when running the app there are no errors or anything. It's just when you relaunch the app you will notice the store is empty. And after a bit of debugging I found that it was caused by an error being thrown in this do catch statement

```
Task { @MainActor in
	do {
			let decoder = JSONDecoder()
			self.items = try await self.storageEngine.readAllData()
					.map({ try decoder.decode(Item.self, from: $0) })
			print(self.items)
	} catch {
			self.items = []
	}
}
```
That also explains why the store is empty after a restart because if an error is thrown, the items will be empty.

**Reply by mergesort** [reply]
Thanks a lot for confirming @batjoc, I should have time to look over this today! Could I ask you to share the code you're testing this with? Most helpful would be the object you're trying to save so I can try to recreate this, but all of the initialized objects and call-sites would be really valuable to me as well.

**Reply by batjoc** [reply]
Sure, see here: https://github.com/batjoc/Boutique/tree/stored-value-bug. I've added a `storedItem` to the `RichNotesController`. Each time you add objects, the first one will be set to the storedItem. Now when you restart the app you will notice that it says there are 0 stored objects.

**Reply by mergesort** [reply]
Hey @batjoc, I took a look and I'm pretty sure you don't have a fresh store, but why that's the case isn't intuitive and is my fault.

If you don't provide a `StorageEngine` parameter to both `@Stored` and `@StoredValue` they will default to `SQLiteStorageEngine(directory: .documents(appendingPath: "Data"))!`. In the buggy demo variant of the repo you provided neither the `Store` in `App.Store` nor the `StoredItem` provide a parameter, so what happens is that they're sharing the same underlying database. When you try to decode a value for one or the other, it's going to encounter a value that wasn't saved with the same schema (given `@Stored` and `@StoredValue` store their data differently), and thus a key decoding error.

If you change this
```swift
extension Store where Item == RichNote {
    static let notesStore = Store<RichNote>(cacheIdentifier: \.id)
}
```

To this
```swift
extension Store where Item == RichNote {
    static let notesStore = Store<RichNote>(
        storage: SQLiteStorageEngine(directory: .defaultStorageDirectory(appendingPath: "Notes"))!,
        cacheIdentifier: \.id
    )
}
```

Then you'll start to see decoding work again once more once you delete your app (since the database is already corrupted).

---
Now of course this isn't something you should have to worry about, and an immediate fix is to make `@StoredValue`'s default parameter be different than `@Stored`'s default parameter. But the best way I can think of to do this is to have an initializer for `@StoredValue` that computes a deterministic hash by default (so you don't have to think about StorageEngines at all), and allow you to still provide a parameter if that's what you think is best for your needs.

I'm not sold on anything though and very open to ideas if anyone has suggestions.

**Reply by batjoc** [reply]
Thanks so much for this @mergesort. I indeed wasn't aware that each store needs a different database. After using core data for such a long time where everything is in one database this never crossed my mind sorry. Thanks again for your great support and this nice new feature in version 2.0!

**Reply by mergesort** [reply]
Of course, just glad I can help! I've put together a pull request for all of v2.0 (#25) including some updates to `@StoredValue` and examples of how to use them, the `key` parameter should now prevent the issue you ran into, plus I added the ability to set default values today. 🙂

There are some merge conflicts I need to address and I'll be making a few more small changes, but I expect none of them _should_ affect your usage, so I think the 2.0 branch is safe for your needs.
